crates/distribution-types/src/uv_requirement.rs

```diff
@@ -0,0 +1,157 @@
+        if let Some(marker) = &self.marker {
+            marker.evaluate(env, extras)
+        } else {
+            true
+        }
+    }
+
+    pub fn from_requirement(requirement: Requirement) -> Result<Self, ParsedUrlError> {
+        let source = match requirement.version_or_url {
+            None => UvSource::Registry {
+                version: VersionSpecifiers::empty(),
+                index: None,
+            },
+            // The most popular case: Just a name, a version range and maybe extras.
+            Some(VersionOrUrl::VersionSpecifier(version)) => UvSource::Registry {
+                version,
+                index: None,
+            },
+            Some(VersionOrUrl::Url(url)) => {
+                let direct_url = ParsedUrl::try_from(&url.to_url())?;
+                UvSource::from_parsed_url(direct_url, url)
+            }
+        };
+        Ok(UvRequirement {
+            name: requirement.name,
+            extras: requirement.extras,
+            marker: requirement.marker,
+            source,
+        })
+    }
+}
+
+impl Display for UvRequirement {
+    /// Note: This is for user display, not for requirements.txt
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.name)?;
+        if !self.extras.is_empty() {
+            write!(
+                f,
+                "[{}]",
+                self.extras
+                    .iter()
+                    .map(ToString::to_string)
+                    .collect::<Vec<_>>()
+                    .join(",")
+            )?;
+        }
+        match &self.source {
+            UvSource::Registry { version, index } => {
+                write!(f, "{version}")?;
+                if let Some(index) = index {
+                    write!(f, " (index: {index})")?;
+                }
+            }
+            UvSource::Url { url, .. } => {
+                write!(f, " @ {url}")?;
+            }
+            UvSource::Git {
+                url: _,
+                repository,
+                reference,
+                subdirectory,
+            } => {
+                write!(f, " @ git+{repository}")?;
+                if let Some(reference) = reference.as_str() {
+                    write!(f, "@{reference}")?;
+                }
+                if let Some(subdirectory) = subdirectory {
+                    writeln!(f, "#subdirectory={}", subdirectory.display())?;
+                }
+            }
+            UvSource::Path { url, .. } => {
+                write!(f, " @ {url}")?;
+            }
+        }
+        if let Some(marker) = &self.marker {
+            write!(f, " ; {marker}")?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Hash, Debug, Clone, Eq, PartialEq)]
+pub enum UvSource {
+    Registry {
+        version: VersionSpecifiers,
+        index: Option<String>,
+    },
+    // TODO(konsti): Track and verify version specifier from pyproject.toml
```

Comment:
What's the work involved here? (What's the goal?)

Reply:
You can do things like `flask==3.0.0` and `flask = { git = "https://github.com/pallets/flask", tag = "v2.0.0" }`. We should error or warn when that happens, it's another consistency validation check.

Reply:
I see, thanks.